core/types.go

```diff
@@ -48,3 +54,47 @@ func (ch Challenge) ExpectedKeyAuthorization(key *jose.JSONWebKey) string {
 
 	return ch.Token + "." + base64.RawURLEncoding.EncodeToString(thumbprint)
 }
+
+type Certificate struct {
+	ID     string
+	Cert   *x509.Certificate
+	DER    []byte
+	Issuer *Certificate
+}
+
+func (c Certificate) PEM() []byte {
+	var buf bytes.Buffer
+
+	err := pem.Encode(&buf, &pem.Block{
+		Type:  "CERTIFICATE",
+		Bytes: c.DER,
+	})
+	if err != nil {
+		panic(fmt.Sprintf("Unable to encode certificate %q to PEM: %s",
+			c.ID, err.Error()))
+	}
+
+	return buf.Bytes()
+}
+
+func (c Certificate) Chain() []byte {
+	chain := make([][]byte, 0)
+
+	// Add the leaf certificate
+	chain = append(chain, c.PEM())
+
+	// Add zero or more issuers
+	issuer := c.Issuer
+	for {
+		// if the issuer is nil, or the issuer's issuer is nil then it isn't a leaf
+		// or an intermediate. Don't put it in the chain
+		if issuer == nil || issuer.Issuer == nil {
```

Comment:
Oops! Spoke too soon. I remember why I did this: without checking `issuer.Issuer == nil` then the root issuer is included in the chain as well as the leaf & intermediate. Removing the check results in a chain of three certs when it should be two.